Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,427 @@
+
+    m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::forward<Args>(args)...);
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      *reinterpret_cast<U**>(m_data.data()) = heap;
+    }
+  }
+
+#if defined(_ENABLE_ACTS_ANY_DEBUG)
+  AnyBase() { _ACTS_ANY_DEBUG("Default construct this=" << this); };
+#else
+  AnyBase() = default;
+#endif
+
+  template <typename T>
+  explicit AnyBase(T&& value) {
+    using U = std::decay_t<T>;
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::move(value));
+      _ACTS_ANY_DEBUG(
+          "Construct local (this=" << this << ") at: " << (void*)m_data.data());
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::move(value));
+      _ACTS_ANY_DEBUG("Construct heap (this=" << this << ") at: " << heap);
+      _ACTS_ANY_DEBUG_BUFFER("-> buffer before", m_data);
+      *reinterpret_cast<U**>(m_data.data()) = heap;
+      _ACTS_ANY_DEBUG_BUFFER("-> buffer after", m_data);
+    }
+  }
+
+  template <typename T>
+  T& as() {
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    assert(makeHandler<T>() == m_handler && "Bad type access");
+    if (m_handler->typeSize <= SIZE) {
+      T* ptr = reinterpret_cast<T*>(m_data.data());
+      _ACTS_ANY_DEBUG("As local: " << ptr);
+      return *ptr;
+    } else {
+      // is heap allocated
+      void* ptr = *reinterpret_cast<void**>(m_data.data());
+      _ACTS_ANY_DEBUG("As heap: " << ptr);
+      return *reinterpret_cast<T*>(ptr);
+    }
+  }
+
+  template <typename T>
+  const T& as() const {
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    assert(makeHandler<T>() == m_handler && "Bad type access");
+    if (m_handler->typeSize <= SIZE) {
+      const T* ptr = reinterpret_cast<const T*>(m_data.data());
+      _ACTS_ANY_DEBUG("As local: " << ptr);
+      return *ptr;
+    } else {
+      // is heap allocated
+      const void* ptr = *reinterpret_cast<void* const*>(m_data.data());
+      _ACTS_ANY_DEBUG("As heap: " << ptr);
+      return *reinterpret_cast<const T*>(ptr);
+    }
+  }
+
+  ~AnyBase() { destroy(); }
+
+  AnyBase(const AnyBase& other) {
+    if (m_handler == nullptr) {  // this object is empty
+      m_handler = other.m_handler;
+      assert(m_handler != nullptr && "Handler is null");
```

Comment:
Why is it considered bad to assign from a null object?

Reply:
This is after assignment, so we assigned an empty to an empty. Doesn't have to be an error, I guess, just a noop.